### PR TITLE
[fix] Buffers, like params go back to fp32 after fwd

### PR DESCRIPTION
**Pull request by sshleifer**
## 🐛 Bug (copy pasted from issue)

In FSDP with `mixed_precision=True`, we only maintain buffers in FP16. This could lead to underflow/overflow for buffers that wouldn't happen with standard AMP training.

Ideally we could keep the buffers in their original dtype, but it's unclear how/when/if to cast to `compute_dtype`. For static buffers it doesn't matter, but for BatchNorm we want updates to happen in full precision.

### CODE REVIEW

**Review by myleott — APPROVED** [review]
Looks good!

Can you also add a test? Maybe add a small buffer to `DeviceAndTypeCheckModule` and then add some extra `expected_buffer_dtype` checks here: https://github.com/facebookresearch/fairscale/blob/master/tests/nn/data_parallel/test_fsdp.py#L153

**Comment by sshleifer** [comment]
Done, asserted that buffers are `param_dtype` during forward and `torch.float32` after forward.
There is no need for `expected_buffer_dtype` given my implem.

**Comment by myleott** [comment]
Ah, please see this comment: https://github.com/facebookresearch/fairscale/issues/442#issuecomment-787553247

I think we want `buffer_dtype` to stay fp32 inside the forward pass even when using mixed precision, whereas param_dtype would be fp16 in this case. This would match the behavior of O2 Apex/AMP: https://nvidia.github.io/apex/amp.html#o2-almost-fp16-mixed-precision

**Review by min-xu-ai — APPROVED** [review]
LGTM

**Comment by sshleifer** [comment]
we went with `buffer_dtype`
